runtime/shared_common/OSCache.cpp

```diff
@@ -739,20 +742,84 @@ SH_OSCache::getAllCacheStatistics(J9JavaVM* vm, const char* ctrlDirName, UDATA g
 			if (includeOldGenerations || !tempInfo.isCompatible || (getCurrentCacheGen() == tempInfo.generation)) {
 				if (tempInfo.isCompatible) {
 					if (!ignoreCompatible) {
-						newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
-						memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+						if (isToplayer) {
+							/* nameWithVGen is a compatible top layer cache, lower layer stats are in lowerLayerList*/
+							newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+							if (NULL == newElement) {
+								Trc_SHR_OSC_getAllCacheStatistics_pool_newElement_failed(0);
+								if (NULL != lowerLayerList) {
+									pool_kill(lowerLayerList);
+								}
+								outOfMem = true;
+								goto done;
+							}
+							memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+							if (NULL != lowerLayerList) {
+								if (0 < pool_numElements(lowerLayerList)) {
+									pool_state poolState;
+									SH_OSCache_Info* anElement = (SH_OSCache_Info*)pool_startDo(lowerLayerList, &poolState);
+									do {
+										newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+										if (NULL == newElement) {
+											Trc_SHR_OSC_getAllCacheStatistics_pool_newElement_failed(1);
+											if (NULL != lowerLayerList) {
+												pool_kill(lowerLayerList);
+											}
+											outOfMem = true;
+											goto done;
+										}
+										memcpy(newElement, anElement, sizeof(SH_OSCache_Info));
+										++cntr;
+									} while ((anElement = (SH_OSCache_Info*)pool_nextDo(&poolState)) != NULL);
+								}
+							}
+						} else {
+							if (SHR_STATS_REASON_ITERATE != reason) {
+								newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+								if (NULL == newElement) {
+									Trc_SHR_OSC_getAllCacheStatistics_pool_newElement_failed(2);
+									if (NULL != lowerLayerList) {
+										pool_kill(lowerLayerList);
+									}
+									outOfMem = true;
+									goto done;
+								}
+								memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+							}
+							/* nameWithVGen is a compatible lower layer cache and getLowerLayerStats is false, do nothing here. stats of non-top layers are included in the lowerLayerList of when calling getCacheStatistics() on the top layer */
+						}
+					} else {
+						/* ignoreCompatible is true, do nothing here */
 					}
 				} else {
 					newElement = (SH_OSCache_Info*) pool_newElement(incompatibleList);
+					if (NULL == newElement) {
+						Trc_SHR_OSC_getAllCacheStatistics_pool_newElement_failed(3);
+						if (NULL != lowerLayerList) {
+							pool_kill(lowerLayerList);
+						}
+						outOfMem = true;
+						goto done;
+					}
 					memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
 				}
 			}
+			++cntr;
 		}
-		++cntr;
+		
+		if (NULL != lowerLayerList) {
```

Comment:
This needs to occur after `done:` to ensure the pool is always killed.

Reply:
Added `pool_kill(lowerLayerList)` after `done:` We still need to clear lowerLayerList in the successful case for each run in this while loop, so this code is kept here.